### ISSUE TITLE
Blazor - Rest API not working in Azure

### ISSUE THREAD

**Issue by sbuchok** [aspnet/AspNetCore#13630]
I've created a Hosted Blazor app (Client Side Blazor, .Net Core Controller backend). Everything runs on my machine (guaranteed) ... obviously, just like it always would ;) But I deploy it to Azure and none of my controller endpoints work. Even for a "GET" it tries to return the index page rather than calling the appropriate endpoint.

Is there a setting in Azure that I need to enable? I have added the extensions for .Net Core 3.0. Where should I be looking to fix this? 

Also, while a "GET" request returns the index page, a "POST" is returning a 405 Method Not Allowed. I've tried modifying the web.config to remove WebDav, but that is doing nothing. 

Kinda stuck. Any help would be greatly appreciated.

Thanks in advance.

**Comment by ichensky** [comment]
>>Where should I be looking to fix this?

 https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/blazor/?view=aspnetcore-3.0

**Comment by sbuchok** [comment]
Thanks for the response, but I can already deploy my application. If you go here: http://daringclub.azurewebsites.net/ and click on Signup in the top right, there is a Test button on that page. 

[HttpGet("[action]")]
        public IActionResult Test()
        {
            return Ok("Hello World");
        }

When the button is clicked, you get a 405 Method Not Allowed error. When run on my local machine, all WebAPI calls work fine. When deployed to Azure, all WebAPI calls return this 405 Method not allowed error.

**Comment by ichensky** [comment]
>>  there is a Test button on that page.
>> [HttpGet("[action]")]
>> 405 Method Not Allowed error

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-2.2

**Comment by sbuchok** [comment]
Sorry, but nothing in that link mentions anything about 405 Method Not Allowed. I have already explained that this is working locally perfectly fine. I'm assuming that this should mean that there is something that needs to be done on Azure that I am not doing that needs to be setup or some setting in the Startup.cs that I need to do in order to get this to work.

If you need more information from me, please indicate what information you need.

**Comment by ichensky** [comment]
When you press 'Test' button do you trying to make HTTP POST request? Because this method: 
[HttpGet("[action]")]
public IActionResult Test()
{
return Ok("Hello World");
}

mapped to handle only HTTP GET request.

**Comment by sbuchok** [comment]
Apparently I can't even get the the site now. Need to fix when I get home. The Test button does a Get Request. I noticed that some get requests look like they return the index page for some reason, however this GET request is returning a 405.

Put it this way, "It works locally, without issue". It's not like this is the only web request that I am testing. This was setup so that I could have the simplest method to test again. I have 50+ controller endpoints and they all work locally. No issues. When I deploy to Azure, they return 405 Method Not Allowed, all of them. Therefore, something is not setup properly for Azure. Whether that is something in Setup.cs that needs to be added or something in Azure, I'm not sure. 

I've tried the usual modifying the web.config to remove webdav and add the http verbs to isapi and ... 
I have just noticed that the version of .NetCore I have installed on Azure is newer than the version for my App. So tonight, I'm going to upgrade my app to the newest version of .Net Core and see if that fixes the issue. 
I'm guessing it will, but won't know till I try.

**Comment by ichensky** [comment]
Seems, there are problems with routing, because if you click `Test` button you'll get Blazor Index.html page in responce or my be you deployed only Blazor client app(client side only) without backend part.

**Comment by vertonghenb** [comment]
> Error message:
`WASM: '<' is an invalid start of a value. Path: $ | LineNumber: 0 | BytePositionInLine: 0.`

Can you show some code to hulp us out?

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us, @sbuchok.
This seems to be something related to your code.
While we do our best to look through all the issues filed here, to get a faster response we suggest posting your questions to [StackOverflow](https://www.stackoverflow.com) using the **asp.net-core-blazor** tag.

**Comment by sbuchok** [comment]
I have a feeling it has to do with the fact that my code is Preview7 while Preview8 is installed on Azure for .Net Core 3. It is also good to post these issues here with their resolution so that someone else using Blazor who gets the same issue will be able to hopefully find an answer easier.

**Comment by sbuchok** [comment]
Ok, I have upgraded my project and still the same issue. I now created a new project based off of the templates. I ran it local without modifications and it works. I then publish it and the Get request isn't working. I am now very convinced that it is not my code.

You can see the issue here: http://daringclub.azurewebsites.net/fetchdata

If there is a template and it's always recommended to use Azure, shouldn't it work in Azure without modification?

If there are Azure settings that need to be set, what are they? I've read a lot of documentation, but still can't find a fix. I will post to StackOverflow as well, but I see this as being a bug in the template that it doesn't work out of the box on Azure.

**Comment by ichensky** [comment]
>> If there is a template and it's always recommended to use Azure, shouldn't it work in Azure without modification?

It works even without azure, you can use any windows/linux hosting provider. 
If your code works on your PC, then it should works the same at any another machine where installed .net core. Seems, you not correctly deploy your code.

**Comment by sbuchok** [comment]
@ichensky I believe you are 100% correct. Apparently when I'm deploying, I'm only deploying the Client and Shared projects. So, because the Server project is probably not getting deployed, that would cause a GET request to return the index page and POST request to not be allowed. All the deployment examples I've seen only show either just the Client or Server, but not both in the same solution. So, I think I just need to find an example for that. I tried setting the deployment to self contained, but there is an issue with there being .net standard and deploying it normally gives an error about .net Core 3 not being installed on the server even though I installed it through the extensions. Ultimately, I'm closing this as it's not an issue with the Code or the template. Thanks for your help.
